File: objc/Project4/Project4/ViewController.m
//
//  ViewController.m
//  Project4
//
//  Created by Jinwoo Kim on 12/22/20.
//

#import "ViewController.h"

@implementation ViewController
- (void)setRepresentedObject:(id)representedObject {
    [super setRepresentedObject:representedObject];

    // Update the view, if already loaded.
}

// MARK: - Life Cycle and View Selection Methods
- (void)viewDidLoad {
    [super viewDidLoad];

    // 1: Create the stack view and add it to our view
    self.rows = [NSStackView new];
    self.rows.orientation = NSUserInterfaceLayoutOrientationVertical;
    self.rows.distribution = NSStackViewDistributionFillEqually;
    self.rows.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.rows];
    
    // 2: Create Auto Layout constaints that pin the stack view to the edges of its container
    [[self.rows.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor] setActive:YES];
    [[self.rows.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor] setActive:YES];
    [[self.rows.topAnchor constraintEqualToAnchor:self.view.topAnchor] setActive:YES];
    [[self.rows.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor] setActive:YES];
    
    // 3: Create an initial column that contains a single web view
    NSView *webView = [self makeWebView];
    NSStackView *column = [NSStackView stackViewWithViews:@[webView]];
    column.distribution = NSStackViewDistributionFillEqually;
    
    // 4: Add this column to the `rows` stack view
    [self.rows addArrangedSubview:column];
}

- (NSView *)makeWebView {
    WKWebView *webView = [WKWebView new];
    NSURL *url = [NSURL URLWithString:@"https://www.apple.com"];
    NSURLRequest *request = [NSURLRequest requestWithURL:url];
    NSClickGestureRecognizer *recognizer = [[NSClickGestureRecognizer alloc] initWithTarget:self action:@selector(webViewClicked:)];
    
    webView.navigationDelegate = self;
    webView.wantsLayer = YES;
    [webView loadRequest:request];
    
    [webView addObserver:self forKeyPath:@"estimatedProgress" options:NSKeyValueObservingOptionNew context:nil];
    
    recognizer.numberOfClicksRequired = 2;
    recognizer.delegate = self;
    [webView addGestureRecognizer:recognizer];
    
    if (self.selectedWebView == nil) [self select:webView];
    
    return webView;
}

- (void)webViewClicked:(NSClickGestureRecognizer *)recognizer {
    // get the web view that triggered this method
    WKWebView *newSelectedWebView = (WKWebView *)recognizer.view;
    if (newSelectedWebView == nil) return;
    
    // deselect the currently selected webview if there is one
    self.selectedWebView.layer.borderWidth = 0;
    
    // select the new one
    [self select:newSelectedWebView];
    
}

- (void)select:(WKWebView *)webView {
    self.selectedWebView = webView;
    self.selectedWebView.layer.borderWidth = 4;
    self.selectedWebView.layer.borderColor = NSColor.blueColor.CGColor;
    
    WindowController *windowController = self.view.window.windowController;
    
    if (windowController != nil) {
        windowController.addressEntry.stringValue = self.selectedWebView.URL.absoluteString;
    }
}

- (void)webView:(WKWebView *)webView didCommitNavigation:(WKNavigation *)navigation {
    if (self.selectedWebView == nil) return;
    
    WindowController *windowController = self.view.window.windowController;
    if (windowController != nil) {
        windowController.addressEntry.stringValue = webView.URL.absoluteString;
    }
}

-(BOOL)gestureRecognizer:(NSGestureRecognizer *)gestureRecognizer shouldAttemptToRecognizeWithEvent:(NSEvent *)event {
    if (gestureRecognizer.view == self.selectedWebView) {
        return NO;
    }
    return YES;
}

// MARK: - Touch Bar
- (void)selectAddressEntry {
    WindowController *windowController = self.view.window.windowController;
    if (windowController == nil) return;
    [windowController.window makeFirstResponder:windowController.addressEntry];
}

- (nonnull NSArray *)itemsForSharingServicePickerTouchBarItem:(nonnull NSSharingServicePickerTouchBarItem *)pickerTouchBarItem {
    if (self.selectedWebView == nil) return @[];
    NSString *url = self.selectedWebView.URL.absoluteString;
    if (url == nil) return @[];
    return @[url];
}

- (NSTouchBar *)makeTouchBar {
    // enable the Customize Touch Bar menu item
    NSApp.automaticCustomizeTouchBarMenuItemEnabled = YES;
    
    // create a Touch Bar with a unique identifier, making `ViewController` its delegate
    NSTouchBar *touchBar = [NSTouchBar new];
    touchBar.customizationIdentifier = NSTouchBarCustomizationIdentifierProject;
    touchBar.delegate = self;
    
    // set up some meaningful defaults
    touchBar.defaultItemIdentifiers = @[
        NSTouchBarItemIdentifierNavigation,
        NSTouchBarItemIdentifierAdjustGrid,
        NSTouchBarItemIdentifierEnterAddress,
        NSTouchBarItemIdentifierSharingPicker
    ];
    
    // make the address entry button sit in the center of the bar
    touchBar.principalItemIdentifier = NSTouchBarItemIdentifierEnterAddress;
    
    // allow the user to customize these four controls
    touchBar.customizationAllowedItemIdentifiers = @[
        NSTouchBarItemIdentifierSharingPicker,
        NSTouchBarItemIdentifierAdjustGrid,
        NSTouchBarItemIdentifierAdjustCols,
        NSTouchBarItemIdentifierAdjustRows
    ];
    
    // but don't let them take off the URL entry button
    touchBar.customizationRequiredItemIdentifiers = @[NSTouchBarItemIdentifierEnterAddress];
    
    return touchBar;
}

- (NSTouchBarItem *)touchBar:(NSTouchBar *)touchBar makeItemForIdentifier:(NSTouchBarItemIdentifier)identifier {
    if (identifier == NSTouchBarItemIdentifierEnterAddress) {
        NSButton *button = [NSButton buttonWithTitle:@"Enter a URL" target:self action:@selector(selectAddressEntry)];
        NSCustomTouchBarItem *customTouchBarItem = [[NSCustomTouchBarItem alloc] initWithIdentifier:identifier];
        
        [button setContentHuggingPriority:10 forOrientation:NSLayoutConstraintOrientationHorizontal];
        customTouchBarItem.view = button;
        return customTouchBarItem;
    } else if (identifier == NSTouchBarItemIdentifierNavigation) {
        // load the back and forward images
        NSImage *back = [NSImage imageNamed:NSImageNameTouchBarGoBackTemplate];
        NSImage *forward = [NSImage imageNamed:NSImageNameGoForwardTemplate];
        
        // create a segmented control out of them, calling our `navigationClicked:` method
        NSSegmentedControl *segmentedControl = [NSSegmentedControl
                                                segmentedControlWithImages:@[back, forward]
                                                trackingMode:NSSegmentSwitchTrackingMomentary
                                                target:self
                                                action:@selector(navigationClicked:)];
        
        // wrap that inside a Touch Bar item
        NSCustomTouchBarItem *customTouchBarItem = [[NSCustomTouchBarItem alloc] initWithIdentifier:identifier];
        customTouchBarItem.view = segmentedControl;
        
        // send it back
        return customTouchBarItem;
    } else if (identifier == NSTouchBarItemIdentifierSharingPicker) {
        NSSharingServicePickerTouchBarItem *picker = [[NSSharingServicePickerTouchBarItem alloc] initWithIdentifier:identifier];
        picker.delegate = self;
        return picker;
    } else if (identifier == NSTouchBarItemIdentifierAdjustRows) {
        NSSegmentedControl *control = [NSSegmentedControl
                                       segmentedControlWithLabels:@[@"Add Row", @"Remove Row"]
                                       trackingMode:NSSegmentSwitchTrackingMomentaryAccelerator
                                       target:self
                                       action:@selector(adjustRows:)];
        NSCustomTouchBarItem *customTouchBarItem = [[NSCustomTouchBarItem alloc] initWithIdentifier:identifier];
        customTouchBarItem.customizationLabel = @"Rows";
        customTouchBarItem.view = control;
        return customTouchBarItem;
    } else if (identifier == NSTouchBarItemIdentifierAdjustCols) {
        NSSegmentedControl *control = [NSSegmentedControl
                                       segmentedControlWithLabels:@[@"Add Column", @"Remove Column"]
                                       trackingMode:NSSegmentSwitchTrackingMomentaryAccelerator
                                       target:self
                                       action:@selector(adjustColumns:)];
        NSCustomTouchBarItem *customTouchBarItem = [[NSCustomTouchBarItem alloc] initWithIdentifier:identifier];
        customTouchBarItem.customizationLabel = @"Column";
        customTouchBarItem.view = control;
        return customTouchBarItem;
    } else if (identifier == NSTouchBarItemIdentifierAdjustGrid) {
        NSPopoverTouchBarItem *popover = [[NSPopoverTouchBarItem alloc] initWithIdentifier:identifier];
        popover.collapsedRepresentationLabel = @"Grid";
        popover.customizationLabel = @"Adjust Grid";
        popover.popoverTouchBar = [NSTouchBar new];
        popover.popoverTouchBar.delegate = self;
        popover.popoverTouchBar.defaultItemIdentifiers = @[NSTouchBarItemIdentifierAdjustRows, NSTouchBarItemIdentifierAdjustCols];
        return popover;
    }
    return nil;
}

// MARK: - IBOAction
- (IBAction)urlEnterted:(NSTextField *)sender {
    // bail out if we don't have a webview selected
    if (self.selectedWebView == nil) return;
    
    // attempt to convert the user's text into a URL
    NSURL *url = [NSURL URLWithString:sender.stringValue];
    if (url == nil) return;
    NSURLRequest *request = [[NSURLRequest alloc] initWithURL:url];
    [self.selectedWebView loadRequest:request];
}

- (IBAction)navigationClicked:(NSSegmentedControl *)sender {
    // make sure we have a webview selected
    if (self.selectedWebView == nil) return;
    
    if (sender.selectedSegment == 0) [self.selectedWebView goBack];
    else [self.selectedWebView goForward];
}

- (IBAction)adjustRows:(NSSegmentedControl *)sender {
    if (sender.selectedSegment == 0) {
        // we're adding a new row
        // count how many columns we have so far
        NSStackView *firstStackView = (NSStackView *)self.rows.arrangedSubviews.firstObject;
        NSUInteger columnCount = firstStackView.arrangedSubviews.count;
        
        // make a new array of webviews that contain the correct number of columns
        NSMutableArray<NSView *> *viewArray = [@[] mutableCopy];
        for (NSUInteger i = 0; i < columnCount; i++) {
            [viewArray addObject:[self makeWebView]];
        }
        
        // use that web view array to create a new stackview
        NSStackView *row = [NSStackView stackViewWithViews:viewArray];
        
        // make the stack view size its children equally, then add it to our `rows` array
        row.distribution = NSStackViewDistributionFillEqually;
        [self.rows addArrangedSubview:row];
    } else {
        // we're deleting a row
        // make sure we have at least two rows
        if (self.rows.arrangedSubviews.count <= 1) return;
        
        // pull out the final row, and make sure it's a stackview
        NSStackView *rowToRemove = (NSStackView *)self.rows.arrangedSubviews.lastObject;
        if (rowToRemove == nil) return;
        
        // loop through each webview in the row, removing it form the screen
        for (NSView *cell in rowToRemove.arrangedSubviews) {
            [cell removeFromSuperview];
        }
        
        // finally, remove the whole stackview row
        [self.rows removeArrangedSubview:rowToRemove];
    }
}

- (IBAction)adjustColumns:(NSSegmentedControl *)sender {
    if (sender.selectedSegment == 0) {
        // we need to add a column
        for (NSStackView *row in self.rows.arrangedSubviews) {
            if (row == nil) continue;
            [row addArrangedSubview:[self makeWebView]];
        }
    } else {
        // we need to delete a column
        // pull out the first of our rows
        NSStackView *firstRow = (NSStackView *)self.rows.arrangedSubviews.firstObject;
        if (firstRow == nil) return;
        
        // make sure it has at least two columns
        if (firstRow.arrangedSubviews.count <= 1) return;
    
        // if we are still here it means it's safe to delete a column
        for (NSStackView *row in self.rows.arrangedSubviews) {
            if (row == nil) continue;
            NSView *last = row.arrangedSubviews.lastObject;
            if (last == nil) continue;
            [row removeArrangedSubview:last];
            [last removeFromSuperview];
        }
    }
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context {
    WKWebView *webView = (WKWebView *)object;
    if (webView == nil) return;
    NSURL *url = webView.URL;
    if (url == nil) return;
    NSLog(@"The URL %@ loaded %f", url.absoluteString, webView.estimatedProgress);
}
@end

File: swift/Project4/Project4/ViewController.swift
//
//  ViewController.swift
//  Project4
//
//  Created by Jinwoo Kim on 12/22/20.
//

import Cocoa
import WebKit

class ViewController: NSViewController, WKNavigationDelegate, NSGestureRecognizerDelegate, NSTouchBarDelegate, NSSharingServicePickerTouchBarItemDelegate {
    
    var rows: NSStackView!
    var selectedWebView: WKWebView!

    override func viewDidLoad() {
        super.viewDidLoad()

        // 1: Create the stack view and add it to our view
        rows = NSStackView()
        rows.orientation = .vertical
        rows.distribution = .fillEqually
        rows.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(rows)
        
        // 2: Create Auto Layout constraints that pin the stack view to the edges of its container
        rows.leadingAnchor.constraint(equalTo: view.leadingAnchor).isActive = true
        rows.trailingAnchor.constraint(equalTo: view.trailingAnchor).isActive = true
        rows.topAnchor.constraint(equalTo: view.topAnchor).isActive = true
        rows.bottomAnchor.constraint(equalTo: view.bottomAnchor).isActive = true
        
        // 3: Create an initial column that contains a single web view
        let column = NSStackView(views: [makeWebView()])
        column.distribution = .fillEqually
        
        // 4: Add this column to the `rows` stack view
        rows.addArrangedSubview(column)
    }

    override var representedObject: Any? {
        didSet {
        // Update the view, if already loaded.
        }
    }
    
    func select(webView: WKWebView) {
        selectedWebView = webView
        selectedWebView.layer?.borderWidth = 4
        selectedWebView.layer?.borderColor = NSColor.blue.cgColor
        
        if let windowController = view.window?.windowController as? WindowController {
            windowController.addressEntry.stringValue = selectedWebView.url?.absoluteString ?? ""
        }
    }
    
    @objc func webViewClicked(recognizer: NSClickGestureRecognizer) {
        // get the web view that triggered this method
        guard let newSelectedWebView = recognizer.view as? WKWebView else { return }
        
        // deselect the currently selected web view if there is one
        if let selected = selectedWebView {
            selected.layer?.borderWidth = 0
        }
        
        // select the new one
        select(webView: newSelectedWebView)
    }
    
    func makeWebView() -> NSView {
        let webView = WKWebView()
        webView.navigationDelegate = self
        webView.wantsLayer = true
        webView.load(URLRequest(url: URL(string: "https://www.apple.com")!))
        
        webView.addObserver(self, forKeyPath: #keyPath(WKWebView.estimatedProgress), options: .new, context: nil)
        
        let recognizer = NSClickGestureRecognizer(target: self, action: #selector(webViewClicked(recognizer:)))
        recognizer.numberOfClicksRequired = 2
        recognizer.delegate = self
        webView.addGestureRecognizer(recognizer)
        
        if selectedWebView == nil {
            select(webView: webView)
        }
        
        return webView
    }
    
    func webView(_ webView: WKWebView, didCommit navigation: WKNavigation!) {
        guard webView == selectedWebView else { return }
        
        if let windowController = view.window?.windowController as? WindowController {
            windowController.addressEntry.stringValue = webView.url?.absoluteString ?? ""
        }
    }
    
    func gestureRecognizer(_ gestureRecognizer: NSGestureRecognizer, shouldAttemptToRecognizeWith event: NSEvent) -> Bool {
        if gestureRecognizer.view == selectedWebView {
            return false
        } else {
            return true
        }
    }
    
    func touchBar(_ touchBar: NSTouchBar, makeItemForIdentifier identifier: NSTouchBarItem.Identifier) -> NSTouchBarItem? {
        switch identifier {
        case NSTouchBarItem.Identifier.enterAddress:
            let button = NSButton(title: "Enter a URL", target: self, action: #selector(selectAddressEntry))
            button.setContentHuggingPriority(NSLayoutConstraint.Priority(10), for: .horizontal)
            let customTouchBarItem = NSCustomTouchBarItem(identifier: identifier)
            customTouchBarItem.view = button
            return customTouchBarItem
        case NSTouchBarItem.Identifier.navigation:
            // load the back and forward images
            let back = NSImage(named: NSImage.touchBarGoBackTemplateName)!
            let forward = NSImage(named: NSImage.touchBarGoForwardTemplateName)!
            
            // create a segmented control out of them, calling our `navigatoinClicked()` method
            let segmentedControl = NSSegmentedControl(images: [back, forward], trackingMode: .momentary, target: self, action: #selector(navigationClicked(_:)))
            
            // wrap that inside a Touch Bar item
            let customTouchBarItem = NSCustomTouchBarItem(identifier: identifier)
            customTouchBarItem.view = segmentedControl
            
            // send it back
            return customTouchBarItem
        case NSTouchBarItem.Identifier.sharingPicker:
            let picker = NSSharingServicePickerTouchBarItem(identifier: identifier)
            picker.delegate = self
            return picker
        case NSTouchBarItem.Identifier.adjustRows:
            let control = NSSegmentedControl(labels: ["Add Row", "Remove Row"], trackingMode: .momentaryAccelerator, target: self, action: #selector(adjustRows(_:)))
            let customTouchBarItem = NSCustomTouchBarItem(identifier: identifier)
            customTouchBarItem.customizationLabel = "Rows"
            customTouchBarItem.view = control
            return customTouchBarItem
        case NSTouchBarItem.Identifier.adjustCols:
            let control = NSSegmentedControl(labels: ["Add Column", "Remove Column"], trackingMode: .momentaryAccelerator, target: self, action: #selector(adjustColumns(_:)))
            let customTouchBarItem = NSCustomTouchBarItem(identifier: identifier)
            customTouchBarItem.customizationLabel = "Columns"
            customTouchBarItem.view = control
            return customTouchBarItem
        case NSTouchBarItem.Identifier.adjustGrid:
            let popover = NSPopoverTouchBarItem(identifier: identifier)
            popover.collapsedRepresentationLabel = "Grid"
            popover.customizationLabel = "Adjust Grid"
            popover.popoverTouchBar = NSTouchBar()
            popover.popoverTouchBar.delegate = self
            popover.popoverTouchBar.defaultItemIdentifiers = [.adjustRows, .adjustCols]
            return popover
        default:
            return nil
        }
    }
    
    override func makeTouchBar() -> NSTouchBar? {
        // enable the Customize Touch Bar menu item
        NSApp.isAutomaticCustomizeTouchBarMenuItemEnabled = true
        
        // create a Touch Bar with a unique identifier, making `ViewController` its delegate
        let touchBar = NSTouchBar()
        touchBar.customizationIdentifier = NSTouchBar.CustomizationIdentifier("com.hackingwithswift.project4")
        touchBar.delegate = self
        
        // set up some meaningful defaults
        touchBar.defaultItemIdentifiers = [.navigation, .adjustGrid, .enterAddress, .sharingPicker]
        
        // make the address entry button sit in the center of the bar
        touchBar.principalItemIdentifier = .enterAddress
        
        // allow the user to customize these four controls
        touchBar.customizationAllowedItemIdentifiers = [.sharingPicker, .adjustGrid, .adjustCols, .adjustRows]
        
        // but don't let them take off the URL entry button
        touchBar.customizationRequiredItemIdentifiers = [.enterAddress]
        
        return touchBar
    }
    
    @objc func selectAddressEntry() {
        if let windowController = view.window?.windowController as? WindowController {
            windowController.window?.makeFirstResponder(windowController.addressEntry)
        }
    }
    
    func items(for pickerTouchBarItem: NSSharingServicePickerTouchBarItem) -> [Any] {
        guard let webView = selectedWebView else { return [] }
        guard let url = webView.url?.absoluteString else { return [] }
        return [url]
    }

    @IBAction func urlEntered(_ sender: NSTextField) {
        // bail out if we don't have a web view selected
        guard let selected = selectedWebView else { return }
        
        // attempt to convert the user's text into a URL
        if let url = URL(string: sender.stringValue) {
            // it worked - load it!
            selected.load(URLRequest(url: url))
        }
    }
    
    @IBAction func navigationClicked(_ sender: NSSegmentedControl) {
        // make sure we have a web view selected
        guard let selected = selectedWebView else { return }
        
        if sender.selectedSegment == 0 {
            // back was tapped
            selected.goBack()
        } else {
            // forward was tapped
            selected.goForward()
        }
    }
    
    @IBAction func adjustRows(_ sender: NSSegmentedControl) {
        if sender.selectedSegment == 0 {
            // we're adding anew row
            // count how many columns we have so far
            let columnCount = (rows.arrangedSubviews[0] as! NSStackView).arrangedSubviews.count
            
            // make a new array of webviews that contain the correct number of columns
            let viewArray = (0..<columnCount).map { _ in makeWebView() }
            
            // use that web view array to create a new stack view
            let row = NSStackView(views: viewArray)
            
            // make the stack view size its children equally, then add it to our `rows` array
            row.distribution = .fillEqually
            rows.addArrangedSubview(row)
        } else {
            // we're deleting a row
            // make sure we have at least two rows
            guard rows.arrangedSubviews.count > 1 else { return }
            
            // pull out the final row, and make sure it's a stack view
            guard let rowToRemove = rows.arrangedSubviews.last as? NSStackView else { return }
            
            // loop through each web view in the row, removing it from the screen
            for cell in rowToRemove.arrangedSubviews {
                cell.removeFromSuperview()
            }
            
            // finally, remove the whole stack view row
            rows.removeArrangedSubview(rowToRemove)
        }
    }
    
    @IBAction func adjustColumns(_ sender: NSSegmentedControl) {
        if sender.selectedSegment == 0 {
            // we need to add a column
            for case let row as NSStackView in rows.arrangedSubviews {
                // loop over each row and add a new web view to it
                row.addArrangedSubview(makeWebView())
            }
        } else {
            // we need to delete a column
            // pull out the first of our rows
            guard let firstRow = rows.arrangedSubviews.first as? NSStackView else { return }
            
            // make sure it has at least two columns
            guard firstRow.arrangedSubviews.count > 1 else { return }
            
            // if we are still here it means it's safe to delete a column
            for case let row as NSStackView in rows.arrangedSubviews {
                // loop over every row
                if let last = row.arrangedSubviews.last {
                    // pull out the last web view in this column and remove it using two-step process
                    row.removeArrangedSubview(last)
                    last.removeFromSuperview()
                }
            }
        }
    }
    
    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        guard let webView = object as? WKWebView else { return }
        guard let url = webView.url else { return }
        print("The URL \(url) loaded \(webView.estimatedProgress)")
    }
}

extension NSTouchBarItem.Identifier {
    static let navigation = NSTouchBarItem.Identifier("com.hackingwithswift.project4.navigation")
    static let enterAddress = NSTouchBarItem.Identifier("com.hackingwithswift.project4.enterAddress")
    static let sharingPicker = NSTouchBarItem.Identifier("com.hackingwithswift.project4.sharingPicker")
    static let adjustGrid = NSTouchBarItem.Identifier("com.hackingwithswift.project4.adjustGrid")
    static let adjustRows = NSTouchBarItem.Identifier("com.hackingwithswift.project4.adjustRows")
    static let adjustCols = NSTouchBarItem.Identifier("com.hackingwithswift.project4.adjustCols")
}
